test/infrastructure/inmemory/internal/server/api/watch.go

```diff
@@ -0,0 +1,191 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package api
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/emicklei/go-restful/v3"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Event records a lifecycle event for a Kubernetes object.
+type Event struct {
+	Type   watch.EventType `json:"type,omitempty"`
+	Object runtime.Object  `json:"object,omitempty"`
+}
+
+// WatchEventDispatcher dispatches events for a single resourceGroup.
+type WatchEventDispatcher struct {
+	resourceGroup string
+	events        chan (*Event)
+}
+
+// OnCreate dispatches Create events.
+func (m *WatchEventDispatcher) OnCreate(resourceGroup string, o client.Object) {
+	if resourceGroup != m.resourceGroup {
+		return
+	}
+	m.events <- &Event{
+		Type:   watch.Added,
+		Object: o,
+	}
+}
+
+// OnUpdate dispatches Update events.
+func (m *WatchEventDispatcher) OnUpdate(resourceGroup string, _, o client.Object) {
+	if resourceGroup != m.resourceGroup {
+		return
+	}
+	m.events <- &Event{
+		Type:   watch.Modified,
+		Object: o,
+	}
+}
+
+// OnDelete dispatches Delete events.
+func (m *WatchEventDispatcher) OnDelete(resourceGroup string, o client.Object) {
+	if resourceGroup != m.resourceGroup {
+		return
+	}
+	m.events <- &Event{
+		Type:   watch.Deleted,
+		Object: o,
+	}
+}
+
+// OnGeneric dispatches Generic events.
+func (m *WatchEventDispatcher) OnGeneric(resourceGroup string, o client.Object) {
+	if resourceGroup != m.resourceGroup {
+		return
+	}
+	m.events <- &Event{
+		Type:   watch.EventType("GENERIC"),
+		Object: o,
+	}
+}
+
+// isWatch is true if the request contains `watch="true"` as a query parameter.
+func isWatch(req *restful.Request) bool {
+	return req.QueryParameter("watch") == "true"
+}
+
+func (h *apiServerHandler) watchForResource(req *restful.Request, resp *restful.Response, resourceGroup string, gvk schema.GroupVersionKind) (reterr error) {
+	ctx := req.Request.Context()
+	queryTimeout := req.QueryParameter("timeoutSeconds")
```

Comment:
>Also because of the other discussion around the timeout, which values is Kubernetes ~ usually setting here?

Kubernetes uses `timeoutSeconds` on the server side from the query paramter. The other discussion is about a client timeout.

Reply:
> which values is Kubernetes ~ usually setting here?

I meant which value client-go is setting here. Was this the 500 seconds (or something) you mentioned?